### DISCUSSION TITLE
tarpaulin causes constant rebuilds

### DISCUSSION THREAD

**Discussion by vmiklos** [xd009642/tarpaulin#772] [Q&A] [answered]
Hi,

First, thanks a lot for making tarpaulin. I was scared that the only decent way to get rust code coverage is via some nightly-only flag and then found this project. :-)

The issue I ran into is: it seems tarpaulin somehow causes unexpected rebuilds. My workflow is that I change the code, then `cargo build`, some manual testing then `cargo tarpaulin --out Lcov` to run the tests + check coverage. Here are the times I get for a small project:

- cargo build: 47,308s, then 0,203s
- cargo tarpaulin --out Lcov: 55,252s, then 3,771s

Is this a usage error on my end? Why does building causes a complete tarpaulin rebuild and vice versa?

My hope would be that once I ran both build and tarpaulin, re-running both of them after a tiny code change would not cause a rebuild of dependencies, etc.

Any ideas?

Thanks,

Miklos

**Comment by xd009642** [comment, marked as answer]
So currently on the latest alpha it was found that there are incremental compilation issues with `-Clink-dead-code` is used meaning that without a clean incremental compilation could cause cargo to hang when building the tests. With 0.16.0 which is the last stable release it had to change the relocation model when linking which I believe internally to the rust compiler also causes a rebuild. 

Some people tackle this by having a separate target directory for tarpaulin built binaries and using the `--target-dir` flag. And then if on the latest alpha you can use `--skip-clean` to then disable the clean.

**Reply by vmiklos** [reply]
Aha, adding `--target-dir $PWD/target-cov` to my `cargo tarpaulin` invocation fixes the problem. Thanks! :-)
